src/main/cpp/faultinj/faultinj.cu

```diff
@@ -0,0 +1,506 @@
+    const std::string injectionTypeKey = "injectionType";
+    const std::string substituteReturnCodeKey = "substituteReturnCode";
+    const std::string percentKey = "percent";
+    const std::string interceptionCountKey = "interceptionCount";
+
+    const int injectionType = (*matchedFaultConfig)
+        .get_optional<int>(injectionTypeKey)
+        .value_or(static_cast<int>(FI_RETURN_VALUE));
+
+    const int substituteReturnCode = (*matchedFaultConfig)
+        .get_optional<int>(substituteReturnCodeKey)
+        .value_or(static_cast<int>(CUDA_SUCCESS));
+
+    const int injectionProbability = (*matchedFaultConfig)
+        .get_optional<int>(percentKey)
+        .value_or(0);
+
+    const int interceptionCount = (*matchedFaultConfig)
+        .get_optional<int>(interceptionCountKey)
+        .value_or(INT_MAX);
+
+    spdlog::trace("considered config domain={} function={} injectionType={} probability={} "
+        "interceptionCount={}",
+        domain, cbInfo->functionName, injectionType, injectionProbability, interceptionCount);
+
+    if (interceptionCount <= 0) {
+        spdlog::trace("skipping interception because hit count reached 0, "
+            "domain={} function={} injectionType={} probability={} interceptionCount={}",
+            domain, cbInfo->functionName, injectionType, injectionProbability, interceptionCount);
+        return;
+    }
+
+    if (injectionProbability < 100) {
+        if (injectionProbability <= 0) {
+            return;
+        }
+        const int rand10000 = rand() % 10000;
+        const int skipThreshold = injectionProbability * 10000 / 100;
+        spdlog::trace("rand1000={} skipThreshold={}", rand10000, skipThreshold);
+        if (rand10000 >= skipThreshold) {
+            return;
+        }
+        spdlog::debug(
+            "matched config based on rand10000={} skipThreshold={} "
+            "domain={} function={} injectionType={} probability={}",
+            rand10000, skipThreshold, domain,
+            cbInfo->functionName, injectionType, injectionProbability);
+    } else {
+        spdlog::debug("matched 100% config domain={} function={} injectionType={} probability={}",
+            domain, cbInfo->functionName, injectionType, injectionProbability);
+    }
+
+    // update counter if not unlimited
+    if (interceptionCount != INT_MAX) {
+        spdlog::debug("updating interception count {}: before locking", interceptionCount);
+        // TODO the lock is too coarse-grained.
+        PTHREAD_CALL(pthread_rwlock_wrlock(&globalControl.configLock));
+        const int interceptionCount = (*matchedFaultConfig).get<int>("interceptionCount");
+        (*matchedFaultConfig).put("interceptionCount", interceptionCount - 1);
+        PTHREAD_CALL(pthread_rwlock_unlock(&globalControl.configLock));
+    }
+
+    switch (injectionType)
+    {
+    case FI_TRAP:
+        faultInjectorKernelTrap<<<1,1>>>();
```

Comment:
Apologies for not catching this sooner: I think we might need a `cudaStreamSynchronize(0)` or something here. It's a tiny kernel.

Reply:
I had a bunch of iterations here https://github.com/NVIDIA/spark-rapids-jni/pull/399/commits/36d2678c590d001f614a42d251b26ce159a7dfcf . Let us do cudaStreamSynchronize then :)